retrieval/targetmanager.go

```diff
@@ -123,78 +154,140 @@ func (tm *TargetManager) Pools() map[string][]*Target {
 
-	mtx sync.RWMutex
+	syncCh          chan struct{}
+	cancelScraping  func()
+	cancelProviders func()
 }
 
-func newScrapeSet(app storage.SampleAppender, cfg *config.ScrapeConfig) *scrapeSet {
-	return &scrapeSet{
-		appender: app,
-		config:   cfg,
-		tgroups:  map[string]map[model.Fingerprint]*Target{},
+func newTargetSet(cfg *config.ScrapeConfig, app storage.SampleAppender) *targetSet {
+	ts := &targetSet{
+		tgroups:    map[string]map[model.Fingerprint]*Target{},
+		scrapePool: newScrapePool(app),
+		syncCh:     make(chan struct{}, 1),
+		config:     cfg,
 	}
+	return ts
 }
 
-// run starts the target providers with the given context and consumes
-// and handles their updates. If the context is done, it blocks until the
-// target scrapers have terminated.
-func (ss *scrapeSet) run(ctx context.Context) {
-	var (
-		providers = providersFromConfig(ss.config)
-		wg        sync.WaitGroup
-	)
+func (ts *targetSet) cancel() {
+	ts.mtx.RLock()
+	defer ts.mtx.RUnlock()
 
-	for name, prov := range providers {
-		var (
-			updates = make(chan []*config.TargetGroup)
-		)
+	if ts.cancelScraping != nil {
+		ts.cancelScraping()
+	}
+	if ts.cancelProviders != nil {
+		ts.cancelProviders()
+	}
+}
+
+func (ts *targetSet) runScraping(ctx context.Context) {
+	ctx, ts.cancelScraping = context.WithCancel(ctx)
+
+	ts.scrapePool.ctx = ctx
+
+Loop:
+	for {
+		// Throttle syncing to once per five seconds.
+		select {
+		case <-ctx.Done():
+			break Loop
+		case <-time.After(5 * time.Second):
+		}
 
+		select {
+		case <-ctx.Done():
+			break Loop
+		case <-ts.syncCh:
+			ts.sync()
+		}
+	}
+
+	// We want to wait for all pending target scrapes to complete though to ensure there'll
+	// be no more storage writes after this point.
+	ts.scrapePool.stop()
+}
+
+func (ts *targetSet) sync() {
+	// TODO(fabxc): temporary simple version. For a deduplicating scrape pool we will
+	// submit a list of all targets.
+	ts.scrapePool.sync(ts.tgroups)
+}
+
+func (ts *targetSet) runProviders(ctx context.Context, providers map[string]TargetProvider) {
+	// Lock for the entire time. This may mean up to 5 seconds until the full initial set
+	// is retrieved and applied.
+	// We could release earlier with some tweaks, but this is easier to reason about.
+	ts.mtx.Lock()
+	defer ts.mtx.Unlock()
+
+	var wg sync.WaitGroup
+
+	if ts.cancelProviders != nil {
+		ts.cancelProviders()
+	}
+	ctx, ts.cancelProviders = context.WithCancel(ctx)
+
+	for name, prov := range providers {
 		wg.Add(1)
-		// The update and stopping operations for the target provider handling are blocking.
-		// Thus the run method only returns if all background processing is complete.
+
+		updates := make(chan []*config.TargetGroup)
+
 		go func(name string, prov TargetProvider) {
-			defer wg.Done()
+			var initial []*config.TargetGroup
+
+			select {
+			case <-ctx.Done():
+				wg.Done()
+				return
+			case initial = <-updates:
+				// First set of all targets the provider knows.
+			case <-time.After(5 * time.Second):
```

Comment:
We have lots of other hard coded timeouts (e.g. in target providers). We should find something reasonable but making them all configurable individually seems to get rather messy.
No reasonable setup is going to have a mix of all different providers. Even 2 dynamic one seems rather rare. I'd argue for having a SD-timeout for all providers and leave it at that.

Anyway, this is not going into master yet. Let's care about that at the end.